### ISSUE TITLE
airbyte-ci connectors test fails intermittently

### ISSUE THREAD

**Issue by bnchrch** [airbytehq/airbyte#31526]
# Overview
**TLDR**: `airbyte-ci connectors test` fails intermittently in CI at the metadata validation step during `pip install`

**Current theory**: Because this is a simple step, only happens sometimes, and only when running against multiple connectors, we think this _might_ be related to Dagger

### 1. Issue reported 
**PR:** https://github.com/airbytehq/airbyte/pull/31329
**summary:** Update to 4 connectors
**symptom:** test failed for 2/4 connectors and the Metadata Validation Step (Destination-weaviate and destiantion-qdrant)
**Error:**
```
creating virtual environment...
creating shared libraries...
upgrading shared libraries...
determining package name from '/airbyte-ci/connectors/metadata_service/lib'...
creating virtual environment...
installing metadata-service from spec '/airbyte-ci/connectors/metadata_service/lib'...
Fatal error from pip prevented installation. Full pip output in file:
    /root/.local/pipx/logs/cmd_2023-10-17_16.50.03_pip_errors.log

pip seemed to fail to build package:
    pydash<7.0.0,>=6.0.2

Some possibly relevant errors from pip install:
    ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE. If you have updated the package versions, please update the hashes. Otherwise, examine the package contents carefully; someone may have tampered with them.
Error installing metadata-service from spec
'/airbyte-ci/connectors/metadata_service/lib'.
```

### 2. Local Investigation
@bnchrch ran the failed connector test locally
```
airbyte-ci connectors --name=destination-weaviate test
```

**Was not able to reproduce: It passed**

### 3. Remote CI investigation
@bnchrch ran the `airbyte-ci` via github actions twice

Once for an individual connector: [GHA RUN HERE](https://github.com/airbytehq/airbyte/actions/runs/6550161196/job/17788668996)
```
airbyte-ci connectors --name=destination-weaviate test
```

and Once for all modified connectors:
* [GHA RUN HERE](https://github.com/airbytehq/airbyte/actions/runs/6550179343/job/17788723654):

```
airbyte-ci connectors --modified test
```

**Was not able to reproduce: Both passed**

### 4. Declare (premature) victory
Since it was working
1. Locally
2. and in ci
3. on the affected branch
4. in both single connector and modified

@bnchrch **said it was fixed** 

### 4. Push commit to trigger tests
This is when we merged latest orgin/master into the original branch to trigger the tests again

**We expected them to pass but they failed just as before**
* [GHA RUN HERE](https://github.com/airbytehq/airbyte/actions/runs/6550318280/job/17789162123)
* [Dagger cloud link](https://dagger.cloud/changebypipelines?filter=dagger.io/git.ref:d1cfdf6799e2fc2ae9c3368729c7022474d630b5)

### 5. Run CI workflow manually (again)
@bnchrch ran`--modified` in CI again:
* [GHA RUN HERE](https://github.com/airbytehq/airbyte/actions/runs/6550421944/job/17789514235)
* [Dagger Cloud Link](https://dagger.cloud/changebypipelines?filter=dagger.io/git.ref:d1cfdf6799e2fc2ae9c3368729c7022474d630b5)

and **Was not able to reproduce: It passed**

### The metadata validate step
Its a very basic step that
1. mounts the connector folder and the metadata_service/lib folder
2. Installs the `metadata_service` tool from metadata_service/lib
3. Runs `metadata_service validate` on the connectors metadata file

There are no cache volumes, no explicit caching of any kind, and the container is not reused

and the error occurs when installing the `metadata_service` folder. (python/poetry project)

**Hierarchy**
If you trace the code you can confirm its simplicity
1. [MetadataValidationStep](https://github.com/airbytehq/airbyte/blob/25dc4f6df2dc58b7930434aeae504b825d43a34f/airbyte-ci/connectors/pipelines/pipelines/pipelines/metadata.py#L22C19-L22C19)
2. [SimpleDockerStep](https://github.com/airbytehq/airbyte/blob/25dc4f6df2dc58b7930434aeae504b825d43a34f/airbyte-ci/connectors/pipelines/pipelines/steps/simple_docker_step.py#L45)
3. [with_installed_pipx_package](https://github.com/airbytehq/airbyte/blob/25dc4f6df2dc58b7930434aeae504b825d43a34f/airbyte-ci/connectors/pipelines/pipelines/actions/environments.py#L119)

### Clue: The connector test report says error, but dagger cloud says pass
I noticed in the last failure case 
1. the [gha logs](https://github.com/airbytehq/airbyte/actions/runs/6550318280/job/17789162123) said destination-weaviate failed
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNEpIQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--d5fdc9e370ec25989e647f7c1a8e0c1a0613f51c/image.png)

2. the [html report](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_allow-openai-compatible-embedding-modes/1697561358/d1cfdf6799e2fc2ae9c3368729c7022474d630b5/destination-weaviate/0.2.2/output.html) said destination-weaviate failed
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNFZIQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--d97725d768d7152604ef88dcf47247e79da63f69/image.png)

3. The [dagger cloud link](https://dagger.cloud/changes/93b9a874-98e1-4cea-97c6-d10d434e3375)said destination-weaviate passed
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNGhIQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--fbecef2d9c955c01fd942c3e0157e336ff67eb31/image.png)

**Which makes me wonder how that could happen**

## Leading Theory
I dont know whats going wrong but I have an inkling that something is not going right at the dagger level.

This is because
1. This works locally, and in CI but seems to struggle during a new commit
2. Im seeing logs on the running not match logs in dagger cloud

## Next steps
Need help from the dagger team!

**Comment by bnchrch** [comment]
Thread with dagger: https://discord.com/channels/707636530424053791/1072668966151004290/1163973816549113908

**Comment by sipsma** [comment]
Interestingly, Dagger's own CI has been getting this same error on-and-off starting ~a few days ago.

E.g. here: https://github.com/dagger/dagger/actions/runs/6552195369/job/17794987073#step:4:1007

It's intermittent too. Also of note is that the run I linked above and other runs where it happened run the tests against dagger engines not hooked up to the cloud cache, so it seems like it can happen w/out that.

---

I just also saw a comment from @helderco on the PR for that run that indicated it could be related to using cache volumes for pip across multiple versions of python: https://github.com/dagger/dagger/pull/5906#issuecomment-1767391090

Which would make sense, we do that in our CI.

If that's the problem, then the solution would be to just include the version of python being used in the name of the cache volume so that there's separate volumes per-version.

@bnchrch does that sound like a plausible explanation given the context of your pipelines?

**Comment by bnchrch** [comment]
@sipsma Thanks for the thorough reply.

It does seem like the issue you linked is the same as ours.

and it does line up with that we use multiple versions of python at airbyte.

Ill update that pip cache to target the correct python version accordingly

**Comment by helderco** [comment]
If you're using a mix of alpine and debian images for example, you should also add a key for that because of the different compilers.

If you look at Python's [platform compatibility tags](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/#platform-compatibility-tags) I'd say there's 3 things you need to include in the cache key if you have different variations on the containers that use the cache volume: python interpreter (CPython, PyPy...), version and platform (can affect compilation from source in some packages, not a problem with wheels).

Assuming they all have the same interpreter, but different versions and base OS:
```python
dagger.CacheVolume("pipcache-py310-slim")
# and
dagger.CacheVolume("pipcache-py311-alpine")
```
Or just use the base image ref to make sure they're compatible:
```python
base = "python:3.10-slim"
dagger.CacheVolume(f"pipcache-{base}")
```
